PetAdoptions/petsite/petsite/Controllers/AdoptionController.cs

```diff
@@ -86,0 +1,96 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text.Json;
+using System.Threading.Tasks;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+
+using PetSite.ViewModels;
+
+
+namespace PetSite.Controllers
+{
+    public class AdoptionController : BaseController
+    {
+        private readonly PetSite.Services.IPetSearchService _petSearchService;
+        private static Variety _variety = new Variety();
+        private readonly ILogger<AdoptionController> _logger;
+
+        public AdoptionController(ILogger<AdoptionController> logger, PetSite.Services.IPetSearchService petSearchService)
+        {
+            _petSearchService = petSearchService;
+            _logger = logger;
+        }
+        
+        // GET: Adoption
+        [HttpGet]
+        public IActionResult Index([FromQuery] Pet pet)
+        {
+            if (EnsureUserId()) return new EmptyResult(); // Redirect happened, stop processing
+            
+            // Check if pet data exists in TempData (from TakeMeHome redirect)
+            if (TempData["SelectedPet"] != null)
+            {
+                var petJson = TempData["SelectedPet"].ToString();
+                pet = JsonSerializer.Deserialize<Pet>(petJson);
+            }
+            
+            return View(pet);
+        }
+        
+
+
+        [HttpPost]
+        public async Task<IActionResult> TakeMeHome([FromForm] SearchParams searchParams)
+        {
+            EnsureUserId();
+            // Add custom span attributes using Activity API (compatible with Application Signals auto-instrumentation)
+            var currentActivity = Activity.Current;
+            if (currentActivity != null)
+            {
+                currentActivity.SetTag("pet.id", searchParams.petid);
+                currentActivity.SetTag("pet.type", searchParams.pettype);
+                currentActivity.SetTag("pet.color", searchParams.petcolor);
+                
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Avoid string interpolation within log messages, as it can blur the lines between variables and the log content. Instead, opt for structured logging, where variables are passed as separate arguments. This improves the clarity of logs and facilitates efficient log analysis. For more information, refer to https://github.com/NLog/NLog/wiki/How-to-use-structured-logging.